valhalla/loki/reach.h

```diff
@@ -3,24 +3,94 @@
 
 #include <valhalla/baldr/directededge.h>
 #include <valhalla/loki/search.h>
+#include <valhalla/thor/dijkstras.h>
 
 constexpr uint8_t kInbound = 1;
 constexpr uint8_t kOutbound = 2;
 
 namespace valhalla {
 namespace loki {
 
+// NOTE: another approach is possible which would still allow for one-at-a-time look up. In this case
+// we could actually keep the tree from the previous expansion and as soon as the tree from the next
+// expansion intersects it we could merge the two and continue. To make that work we'd need to remove
+// the part of the expansion that isn't relevant to the current expansion and re-sort the edge set.
+// That would not be an easy task. Instead we could just use the intersection as a short circuit to
+// terminate the expansion if the threshold has been met. The problem here is one of diminishing
+// returns. Which expansion do you keep around for performing the intersections. Surely not all of
+// them, so the question is which ones. The first one may not be relevant for the second one but may
+// be for the 3rd one. The trick is we dont keep any of them around, instead we just remember the
+// the outcomes for any edges that we touched. Specifically we keep a set of edge ids we found to be
+// reachable. We can use this information to shortcut the current expansion when it intersects these
+// edges. If you are searching outbound and hit an edge that is reachabile outbound then you can stop
+// and claim reachable outbound. The same works for the inbound direction. Sadly you cannot use
+// previously found unreachable edges to shortcut a search as they could be local minima (just a
+// branch) of your current search tree
+
 struct directed_reach {
   uint32_t outbound : 16;
   uint32_t inbound : 16;
 };
 
-directed_reach SimpleReach(const valhalla::baldr::DirectedEdge* edge,
-                           uint32_t max_reach,
-                           valhalla::baldr::GraphReader& reader,
-                           const sif::EdgeFilter& edge_filter,
-                           const sif::NodeFilter& node_filter,
-                           uint8_t direction = kInbound | kOutbound);
+class Reach : public thor::Dijkstras {
+public:
+  Reach();
+  // TODO: currently this interface has no place for time, we need to both add it and handle
+  // TODO: the problem of guessing what time to use at the other end of the route depending on
```

Comment:
Is it even going to be possible to do that for a bidirectional search?

Reply:
its definitely possible the question is how close to correct will it be. its an issue for another PR for sure